### ISSUE TITLE
Combining 'delete' and 'replace' for the same RRSet in one request leads to duplicate RRSet error

### ISSUE THREAD

**Issue by similarbg** [PowerDNS/pdns#8610]
I noticed unexpected behaviour in the PowerDNS v4.2.0 API, when pushing new contents for a zone.

Combining 'delete' and 'replace' for the same RRSet in one request leads to a duplicate RRSet error.
This issue was discussed on IRC with 'Habbie'.

A simple snippet to reproduce this error is:
https://gist.github.com/Habbie/edfafea62231b322ae5f8de96436a50b

Req.3 works, req.2 fails with the error: "ApiError (422): Duplicate RRset aaa1._domainkey.example.com. IN CNAME"

The reason I want to combine the 'delete' and the 'replace' rrsets in one request is to replace the complete contents of zone in an atomic way, without risking some rrsets being leftover from the old state.

**Comment by james-stevens** [comment]
Creating duplicate records, especially CNAME records, is clearly an error, but I would have though the whole point of REPLACE is to give you an atomic operation for changing RRs, obviating the need for the DELETE, surley?

**Comment by zeha** [comment]
Yeah, sending delete + replace for the same RRset is generally pointless.

**Comment by similarbg** [comment]
My use case is a situation in which a newly pushed zone might be out of sync with the zone in PowerDNS. Specifically, when a CNAME exists in PowerDNS and an A record is pushed for the same name.
In this case, we need to request the current RRs from PowerDNS, do a smart local diff to determine what RRs should be replaced and what should be deleted (CNAME, because of conflicts with the A RRset) and then call the API with the results. This is not atomic and could lead to race conditions, in case the zone is changed in between these requests.

Ideally we could just push the complete new zone to the PowerDNS API and let PowerDNS atomically replace the old zone with the new zone data. 

This feature does not seem available (yet), so for now I try to simplify the procedure by making it possible to push deletes for all current RRs in PowerDNS and replaces for the new RRs for the zone, so the local diff can be reduced to a basic comparison.

**Comment by zeha** [comment]
You could still just send REPLACE name/CNAME + REPLACE name/A?
I don't see why you need to send a DELETE for name/A and a REPLACE for name/A?

Maybe you can send which data you actually send and why?

**Comment by james-stevens** [comment]
I think what he's saying is they have a situation where they can't be certain what records are held at P/DNS (including possible CNAMEs), but whatever is there needs to be replaced. So they DELETE all rrsets and PATCH in what should be there, in one transaction - but this is leading to a duplicate.

`You could still just send REPLACE name/CNAME + REPLACE name/A?` -> this assumes you already know what data is in pdns.

If they query what is there first, then DELETE it, *or* do the DELETE and PATCH in separate rest calls, there's a race condition. Although, I'd have though, if you DELETE & PATCH (each host) in two API calls, the chances of a problem is slight - but depends on your use case.

Sounds like you're trying to emulate `pdnsutil load-zone <zone> <file>` using the API ... but with no race conditions :)

**Comment by similarbg** [comment]
@james-stevens exactly :)

**Comment by james-stevens** [comment]
But then, surley, you can't avoid having to query pdns for all the _existing_ zone data, becuase you are going to need to remove hosts that exist in pdns, but do **not** exist in the uploaded zone data.

And if you're going to have to query pdns for the existing zone data, then you can use the `REPLACE`, `REPLACE` strategy described above.

Now, whether a `replace` to remove a `CNAME` **and** a `replace` to add an `A` in its place in one transaction will work or not, is another issue. I seems to recall we'd had issues with that, but I can't recall the exact circumstances.

**Comment by similarbg** [comment]
> But then, surley, you can't avoid having to query pdns for all the existing zone data, becuase you are going to need to remove hosts that exist in pdns, but do **not** exist in the uploaded zone data.

It is not relevant whether the data is in the uploaded zone data, because I could just simply delete all RRsets instead of diffing the RRsets locally to determine which records should be deleted and which should be replaced.

In my opinion the current validation preventing the combination of REPLACE and DELETE combination for a  single RRset provides no additional value. 
To be able to combine REPLACE and DELETE would add value, because an incorrect zone configuration in PowerDNS can be easily overwritten without the REPLACE actions triggering validation errors.

**Comment by Habbie** [comment]
Fixed in #8623
